### PR TITLE
Repeating answers for questionnaire item of type Choice using checkbox

### LINKED ISSUE #216: Handle Questionnaire with item repeats to handle questions which can have more than one answer
[Questionnaire.item.repeats](https://www.hl7.org/fhir/questionnaire-definitions.html#Questionnaire.item.repeats) is used to indicate that a particular question / group of questions can have multiple answers. 

> When rendering the questionnaire, it is up to the rendering software whether to render the question text for each answer repetition (i.e. "repeat the question") or to simply allow entry/selection of multiple answers for the question (repeat the answers).
This issue is linked to the following:
- [ ] https://github.com/google/android-fhir/issues/606

### PR DESCRIPTION
**Pull request by joiskash**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #216  partially

**Description**
Adding functionality to handle multiple answer for choice type question

**Type**
Feature

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Review by jingtang10 — COMMENTED** [review]
Thanks @joiskash!

For every type of widget, adding the question value to the questionnaire item is going to be very similar anyway. the variability is in the UI class. so I feel like maybe we can do without the new abstract methods. What do you think? Don't feel like they're adding anything...

**Comment by joiskash** [comment]
> 
> 
> Thanks @joiskash!
> 
> For every type of widget, adding the question value to the questionnaire item is going to be very similar anyway. the variability is in the UI class. so I feel like maybe we can do without the new abstract methods. What do you think? Don't feel like they're adding anything...

Yes I agree, instead of having these abstract method, each UI class can implement its own method to add and remove views for new answers. In the case of checkbox it is trivial because you can just tick as many as you wish.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                }
+              )
+            } else {
+              addAnswer(
+                QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+                  value = answerOption.value
+                }
+              )
+            }
+          }
```
Comment (aditya-07) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt]: Can we extract the view creation logic into a separate function for better readability?

**Review by aditya-07 — CHANGES_REQUESTED** [review]
Added a few comments. Please take a look.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      override fun bind(questionnaireItemViewItem: QuestionnaireItemViewItem) {
+        this.questionnaireItemViewItem = questionnaireItemViewItem
+        if (!questionnaireItemViewItem.questionnaireItem.prefix.isNullOrEmpty()) {
+          prefixTextView.visibility = View.VISIBLE
+          prefixTextView.text = questionnaireItemViewItem.questionnaireItem.localizedPrefix
+        } else {
+          prefixTextView.visibility = View.GONE
+        }
+        val (questionnaireItem, questionnaireResponseItem) = questionnaireItemViewItem
+        val answer = questionnaireResponseItem.answer
```
[the remark this answers is not in the document]
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt:52]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        checkboxGroupHeader.text = questionnaireItem.localizedText
+        checkboxGroup.removeAllViews()
+        var index = 0
```
[the remark this answers is not in the document]
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt:55]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        questionnaireItem.answerOption.forEach { answerOption ->
+          val prefix = TextView(checkboxGroup.context)
+          val checkbox = CheckBox(checkboxGroup.context)
+          val linearLayout = LinearLayout(checkboxGroup.context)
+          if(questionnaireItemViewItem.hasAnswerOption(answerOption)){
+            checkbox.isChecked = true
+          }
```
[the remark this answers is not in the document]
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt (hunk 1 above)
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt]: done

**Comment by joiskash** [comment]
> 
> 
> Added a few comments. Please take a look.

Thanks for the review, I have pushed the changes. Please have a look

**Review by aditya-07 — APPROVED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      ) {
+        questionnaireItemViewItem.addAnswer(questionnaireResponseItemAnswerComponent)
+      }
+
+      fun removeAnswer(
+        questionnaireResponseItemAnswerComponent:
+          QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
+      ) {
+        questionnaireItemViewItem.removeAnswer(questionnaireResponseItemAnswerComponent)
+      }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt:72]: just remove these two?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+      private fun populateViewWithAnswerOption(
+        answerOption: Questionnaire.QuestionnaireItemAnswerOptionComponent
+      ) {
+        val prefix = TextView(checkboxGroup.context)
+        val checkbox = CheckBox(checkboxGroup.context)
+        val linearLayout = LinearLayout(checkboxGroup.context)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt:79]: at this point i think i would just define a layout file for this?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/res/layout/questionnaire_item_checkbox_group_view.xml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    <LinearLayout
+        android:id="@+id/checkbox_group"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="vertical"
+    >
+    </LinearLayout>
+
+
```
[the remark this answers is not in the document]
Comment (joiskash) [reply datacapture/src/main/res/layout/questionnaire_item_checkbox_group_view.xml]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt (hunk 6 above)
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt [hunk 8]
```diff
@@ -49,4 +49,35 @@ internal data class QuestionnaireItemViewItem(
...
+
+  internal fun addAnswer(
+    questionnaireResponseItemAnswerComponent:
+      QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
+  ) {
+    if (!questionnaireItem.repeats) {
+      throw IllegalArgumentException(
+        "Questionnaire item with linkId ${questionnaireItem.linkId} does not allow repeated answers"
+      )
+    }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt:61]:
```suggestion
   check(questionnaireItem.repeats) { "Questionnaire item with linkId ${questionnaireItem.linkId} does not allow repeated answers" }
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt [hunk 9]
```diff
@@ -49,4 +49,35 @@ internal data class QuestionnaireItemViewItem(
...
+
+  internal fun removeAnswer(
+    questionnaireResponseItemAnswerComponent:
+      QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
+  ) {
+    if (!questionnaireItem.repeats) {
+      throw IllegalArgumentException(
+        "Questionnaire item with linkId ${questionnaireItem.linkId} does not allow repeated answers"
+      )
+    }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt:73]:
```suggestion
    check(questionnaireItem.repeats) { "Questionnaire item with linkId ${questionnaireItem.linkId} does not allow repeated answers" }
```

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+          addAnswerOption(
+            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
+              value =
+                Coding().apply {
+                  code = "code 2"
+                  display = "Coding 2"
+                }
+            }
+          )
+        },
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt:167]: just remove this it's covered by the previous test case

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+    checkBox1.performClick()
+    checkBox2.performClick()
+
+    val answer = questionnaireItemViewItem.questionnaireResponseItem.answer
+    assertThat(answer.size).isEqualTo(2)
+    assertThat(answer[0].valueCoding.display).isEqualTo("Coding 1")
+    assertThat(answer[1].valueCoding.display).isEqualTo("Coding 2")
+  }
+}
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt:266]: I thin the test here should be 1) add answer and 2) remove answer.

single or more answers don't really matter and is not what we need to test in my opinion.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by jingtang10** [comment]
can you also change the PR title to something more readable please? thanks

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        }
+      }
+
+      private fun populateViewWithAnswerOption(
+        answerOption: Questionnaire.QuestionnaireItemAnswerOptionComponent
+      ) {
+        val singleCheckBox =
+          LayoutInflater.from(checkboxGroup.context)
+            .inflate(R.layout.questionnaire_item_single_checkbox_view, null)
+        val displayText = singleCheckBox.findViewById<TextView>(R.id.single_check_box_text)
```
[the remark this answers is not in the document]
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt (hunk 8 above)
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt:61]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt (hunk 9 above)
Comment (joiskash) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewItem.kt:73]: done

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt (hunk 10 above)
Comment (joiskash) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt:167]: done

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt (hunk 11 above)
Comment (joiskash) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactoryInstrumentedTest.kt]: done

**Review by jingtang10 — APPROVED** [review]
Thanks @joiskash 👍

**Pull request merged**